File: LeetCode/Explore/Tencent/LinkList/MergeTwoLists.cpp
#include "../../main.h"

class Solution {
  public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        ListNode *p = l1, *q = l2;
        ListNode *head = new ListNode(-1), *k = head;
        while (p != nullptr && q != nullptr) {
            ListNode *t;
            if (p->val < q->val) {
                t = new ListNode(p->val);
                p = p->next;
            } else {
                t = new ListNode(q->val);
                q = q->next;
            }
            t->next = k->next;
            k->next = t;
            k = k->next;
        }
        k->next = p != nullptr ? p : q;
        p = head;
        head = head->next;
        delete p;
        return head;
    }
};

File: LeetCode/Explore/LinkedList/mergeTwoLists.go
package linklist

func mergeTwoLists(l1, l2 *ListNode) *ListNode {
	var head = &ListNode{}
	var p = head
	for l1 != nil && l2 != nil{
		if l1.Val <= l2.Val {
			p.Next = l1
			l1 = l1.Next
		} else {
			p.Next = l2
			l2 = l2.Next
		}
		p = p.Next
	}
	if l1 != nil {
		p.Next = l1
	} else {
		p.Next = l2
	}
	return head.Next
}